pulser-core/pulser/sequence/sequence.py

```diff
@@ -376,74 +377,130 @@ def switch_device(
+            )
         # Initialize the new sequence
         new_seq = Sequence(self.register, new_device)
 
-        # Recall the calls
-        called = False
+        # Channel Matching
+
+        matched = False
         strict_error_message = None
         ch_type_er_mess = None
-        for call in self._calls[1:]:
-            if call.name == "declare_channel":
-                od_ch_obj = self._schedule[call.args[0]].channel_obj
-                for nd_ch_id, nd_ch_obj in new_seq.available_channels.items():
-                    # Find the corresponding channel on the new device
-                    # We verify the channel class then
-                    # check whether it's Global or local
-                    basis_match = od_ch_obj.basis == nd_ch_obj.basis
-                    addressing_match = (
-                        od_ch_obj.addressing == nd_ch_obj.addressing
-                    )
-                    if basis_match and addressing_match:
-                        if strict:
-                            phase_jump_time_check = (
-                                od_ch_obj.phase_jump_time
-                                != nd_ch_obj.phase_jump_time
-                            )
-                            clock_period_check = (
-                                od_ch_obj.clock_period
-                                != nd_ch_obj.clock_period
+        channel_matching = {}
+        sample_seq = sample(self)
+        alert1 = False
+        for channel_name, channel_obj in self.declared_channels.items():
+            od_ch_obj = channel_obj
+            for nd_ch_id, nd_ch_obj in new_seq.available_channels.items():
+                # Find the corresponding channel on the new device
+                # We verify the channel class then
+                # check whether the addressing Global or local
+                basis_match = od_ch_obj.basis == nd_ch_obj.basis
+                addressing_match = od_ch_obj.addressing == nd_ch_obj.addressing
+                if basis_match and addressing_match:
+                    if strict:
+                        ch_samples = sample_seq.channel_samples[channel_name]
+                        ch_sample_phase = ch_samples.phase
+                        # Find if there is phase change between pulses or not
+                        phase_jump_time_oracle = True
+                        if ch_sample_phase.size != 0:
+                            phase_jump_time_oracle = bool(
+                                np.all(ch_sample_phase == ch_sample_phase[0])
                             )
-                            if phase_jump_time_check or clock_period_check:
+                        phase_jump_time_check = (
+                            od_ch_obj.phase_jump_time
+                            == nd_ch_obj.phase_jump_time
+                        )
+                        clock_period_check = (
+                            nd_ch_obj.clock_period % od_ch_obj.clock_period
+                            == 0
```

Comment:
I think this should be the inverse:
```suggestion
                            od_ch_obj.clock_period % nd_ch_obj.clock_period
                            == 0
```
You can check with an example: if `new_clock = 1` and `old_clock = 4`, it should be valid, but `1 % 4 = 1`.